File: Algorithms/Python3.x/41-First_Missing_Positive.py
# Runtime: 36 ms, faster than 83.01% of Python3 online submissions for First Missing Positive.
# Memory Usage: 13.1 MB, less than 65.86% of Python3 online submissions for First Missing Positive.
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        i, n = 0, len(nums)
        while i < n:
            print(nums)
            if 0 < nums[i] <= n and nums[i] != nums[nums[i] - 1]:
                tmp = nums[i]
                nums[i] = nums[nums[i] - 1]
                nums[tmp - 1] = tmp
            else:
                i += 1
        for idx, val in enumerate(nums):
            if val != idx + 1:
                return idx + 1
        return n + 1

File: Algorithms/C++/41-First_Missing_Positive.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for First Missing Positive.
// Memory Usage: 8.7 MB, less than 80.00% of C++ online submissions for First Missing Positive.
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        nums.push_back(0);
        int n = nums.size();
        for (int i = 0; i < nums.size(); ++i)
            if (nums[i] < 0 || nums[i] >= n)
                nums[i] = 0;
        for (int i = 0; i < nums.size(); ++i)
            nums[nums[i] % n] += n;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] / n == 0)
                return i;
        }
        return nums.size();
    }
};
